### DISCUSSION TITLE
Upload Multiple Images

### DISCUSSION THREAD

**Discussion by neilfurry** [symfony/symfony#42975] [Q&A]
Hi,

Im new with symfony and working on a multiple image upload page. it works well on saving images on a directory, however, it doesnt work on saving filename to database. It will just save one of the multiple files i uploaded.

How can  i solve this problem?

Thank you

**Comment by lecneri** [comment]
Can you exemplify your code

**Reply by neilfurry** [reply]
i have this entity

use App\Repository\UploadsRepository;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity(repositoryClass=UploadsRepository::class)
 */
class Uploads
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\Column(type="string", length=255, nullable=true)
     */
    private $path;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $imgName;

    /**
     * @ORM\Column(type="string", length=255)
     */
    private $ftype;

}

and my controller

/**
     * @Route("/admin/upload-photos-and-videos", name="upload_files")
     */
    public function uploadFiles(Request $request): Response
    {
        
        $form = $this->createForm(UploadFilesType::class, null, array(
            'attr' => array(
                'id' => 'upload_form'
            )
        ));
        $form->handleRequest($request);

        if($form->isSubmitted() && $form->isValid())
        {

            $files= $form->get('images')->getData();

                foreach($files as $file){
              

                    $filename = md5(uniqid()) . '.' . $file->guessExtension();
                    $ftype = $file->guessExtension();
                  
                    $file->move(
                        $this->getParameter('uploads_dir'),
                        $filename
                    );
                    
                   
                   
                }

                
                $uploads = new Uploads();
                $saveToDB = $this->getDoctrine()->getManager();
                $uploads->setImgName($filename);
                $uploads->setFtype($ftype);
                $saveToDB->persist($uploads);
                $saveToDB->flush();

                $this->addFlash(
                    'info',
                    'Successfully Uploaded!'
                );

            return $this->redirectToRoute('upload_files');

        }

        return $this->render('admin/upload-files.html.twig', [
            'controller_name' => 'Upload Photos and Videos',
            'uploadFiles' => $form->createView()
        ]);
    }

and here is my controller:

class UploadFilesType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('images', FileType::class, [
                'attr' => [
                    'style' => 'display:none',
                    'multiple' => true,
                    'accept' => 'image/*,video/*'
                ],
                'label' => false,
                'mapped' =>false,
                'multiple' => true
            ])
            ->add('button', ButtonType::class, [
                'attr' => [
                    'class' => 'btn btn-primary mt-5'
                ],
                'label' => 'SELECT PHOTOS/VIDEOS'
            ])
            
        ;
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'data_class' => Uploads::class,
        ]);
    }
}

as i mentioned, uploading photos to directory works fine, the only problem is when i save the image names to database, it only save one from the bunch of images i uploaded.

i hope you can help.

Thank you

**Reply by lecneri** [reply]
Try to change your controller to
```php
$builder
   ->add('images[]', FileType::class, [ //<---- CHANGE THIS
      'attr' => [
         'style' => 'display:none',
         'multiple' => true,
         'accept' => 'image/,video/'
      ],
      'label' => false,
      'mapped' =>false,
      'multiple' => true
])
->add('button', ButtonType::class, [
'attr' => [
'class' => 'btn btn-primary mt-5'
],
'label' => 'SELECT PHOTOS/VIDEOS'
])
```

then dump $request->files to see if there're multiple files.

**Reply by neilfurry** [reply]
this has an illegal character images[], its not valid

**Reply by lecneri** [reply]
you need somehow generate your <input type='file' name='images[]' /> with brakets. I'd go back to collection or generate the input by hand.

**Comment by lecneri** [comment]
Fast looking your code, one thing I suggest you to change, since it's many images, is to modify your UploadFilesType and images become a CollectionType. 
https://symfony.com/doc/current/reference/forms/types/collection.html

**Reply by lecneri** [reply]
Nvm, I see you used option multiple.

**Reply by neilfurry** [reply]
yes, i used multiple options and i see its submitting files to the directory which is correct, but its not saving all the names to database., its just saving one name from the 3 images i uploaded.

**Comment by ker0x** [comment]
It's because you persist only one entity after the `foreach`, you need to move it inside.

```php
$saveToDB = $this->getDoctrine()->getManager();

foreach($files as $file){
    $filename = md5(uniqid()) . '.' . $file->guessExtension();
    $ftype = $file->guessExtension();
              
    $file->move(
        $this->getParameter('uploads_dir'),
        $filename
    );

    $uploads = new Uploads();
    $uploads->setImgName($filename);
    $uploads->setFtype($ftype);

    $saveToDB->persist($uploads);
}

$saveToDB->flush();

```

**Comment by fd6130** [comment]
This might not be an answer but how about giving it a try on using the file upload bundle? It will help you to avoid the needs for writing a file upload function by your own.

https://github.com/dustin10/VichUploaderBundle
